README.md
# @megafon/msisdn-formatter

[![npm (scoped with tag)](https://img.shields.io/npm/v/@megafon/msisdn-formatter/latest?label=%40megafon%2Fmsisdn-formatter)](https://www.npmjs.com/package/@megafon/msisdn-formatter/v/latest) 
[![msisdn CI](https://github.com/MegafonWebLab/msisdn-formatter/actions/workflows/release.yaml/badge.svg)](https://github.com/MegafonWebLab/msisdn-formatter/actions/workflows/release.yaml)

Russian msisdn normalizer and formatter.

## Installation

```
$ yarn add @megafon/msisdn-formatter
```

## Usage

```typescript
import { clean, pretty, PrettyFormats } from '@megafon/msisdn-formatter';

const userNumber = '+79260000000';

console.log(clean(userNumber)); // Prints: '9260000000'
console.log(pretty(userNumber)); // Prints ZETA format by default: '+7 (926) 000-00-00'
console.log(pretty(userNumber, PrettyFormats.ETA)); // Prints: '+7 926 000-00-00'
```

### Available formats:

* alpha: 9161234567
* beta: 79161234567
* gamma: 916 123-4567
* delta: (916) 123-4567
* epsilon: +7 (916) 123-4567
* zeta: +7 (916) 123-45-67
* eta: +7 916 123-45-67

## Contributing

Follow [CONTRIBUTING.md](CONTRIBUTING.md) and [CODE_OF_CONDUCT.md](CODE_OF_CONDUCT.md).


CHANGELOG.md
# [0.4.0](https://github.com/MegafonWebLab/msisdn-formatter/compare/v0.3.0...v0.4.0) (2023-03-22)

### Features

- add customPattern param for the any phone code as allowed ([90ef228](https://github.com/MegafonWebLab/msisdn-formatter/commit/90ef228d4eeca90ddc0efbec5f4db2e9fe453797))

### Other

- Merge pull request #13 from astranavt/15294-allow-8 ([0f807be](https://github.com/MegafonWebLab/msisdn-formatter/commit/0f807be69420b272acbd0a9cc21249c00f7903da)), closes [#13](https://github.com/MegafonWebLab/msisdn-formatter/issues/13)

# [0.3.0](https://github.com/MegafonWebLab/msisdn-formatter/compare/v0.2.0...v0.3.0) (2022-09-16)

### Continuous Integration

- common workflow ([d311a76](https://github.com/MegafonWebLab/msisdn-formatter/commit/d311a7663420d936de5cb804e73c73446bf038d6))

### Documentation

- update README.md ([15913a3](https://github.com/MegafonWebLab/msisdn-formatter/commit/15913a37a156630ceee34f438c2ca096b25865d2))

### Features

- allowed to enter 4 as second phone digit in addittion to 6 and 9 ([24625b9](https://github.com/MegafonWebLab/msisdn-formatter/commit/24625b9dcb3e5fdfd68b21c42f51aedd07c01df0))

### Other

- Merge pull request #10 from Rakot1234/update-phone-pattern ([f0892f6](https://github.com/MegafonWebLab/msisdn-formatter/commit/f0892f6f49aa80c7414d5cad6d6b3be29e41b058)), closes [#10](https://github.com/MegafonWebLab/msisdn-formatter/issues/10)
- Merge pull request #8 from MegafonWebLab/dependabot/npm_and_yarn/ansi-regex-5.0.1 ([bd96474](https://github.com/MegafonWebLab/msisdn-formatter/commit/bd96474b6892605ec47cba70e095f4c0116fc750)), closes [#8](https://github.com/MegafonWebLab/msisdn-formatter/issues/8)
- Bump ansi-regex from 5.0.0 to 5.0.1 ([36d5467](https://github.com/MegafonWebLab/msisdn-formatter/commit/36d5467165007ca90dfa32cc1471ba83e4612786))
- Merge pull request #7 from jekatigr/master ([60cb379](https://github.com/MegafonWebLab/msisdn-formatter/commit/60cb37987648a2f446c0c68bdb7037512fcb59c2)), closes [#7](https://github.com/MegafonWebLab/msisdn-formatter/issues/7)
- Merge pull request #6 from MegafonWebLab/dependabot/npm_and_yarn/tmpl-1.0.5 ([e65d9a2](https://github.com/MegafonWebLab/msisdn-formatter/commit/e65d9a2a3c1b8a0dd689b8d012e32c63e60dc6c0)), closes [#6](https://github.com/MegafonWebLab/msisdn-formatter/issues/6)
- Bump tmpl from 1.0.4 to 1.0.5 ([5ef4383](https://github.com/MegafonWebLab/msisdn-formatter/commit/5ef438373c59af79a66d447effa148ab82020c4c))
- Update SECURITY.md ([6136486](https://github.com/MegafonWebLab/msisdn-formatter/commit/61364863ab7a01478499da680f140742527bf268))
- Merge pull request #5 from IggyStooges/update-readme-by-default ([6e2e49a](https://github.com/MegafonWebLab/msisdn-formatter/commit/6e2e49a67552cb87b7bd6dca915f6a9f0175dfea)), closes [#5](https://github.com/MegafonWebLab/msisdn-formatter/issues/5)
- update readme ([39b8b43](https://github.com/MegafonWebLab/msisdn-formatter/commit/39b8b433cd8914903138c9f54110a6c3eac9ef93))


src/index.test.ts
import { clean, pretty, PrettyFormats } from './index';

describe('msisdn formatter', () => {
    describe('pretty correct cases', () => {
        const testMsisdns = [
            '89161 23 45-67',
            '+7 916 123-4567',
            '79161234567',
            '9161234567',
            '916 123-4567',
            '(916) 123-4567',
            '+7 (916) 123-4567',
            89161234567,
            79161234567,
            9161234567
        ];

        type TestCase = [string | number, PrettyFormats | string | undefined, string];
        const testTable: TestCase[] = [
            ...testMsisdns.map((t): TestCase => [t, PrettyFormats.ALPHA, '9161234567']),
            ...testMsisdns.map((t): TestCase => [t, PrettyFormats.BETA, '79161234567']),
            ...testMsisdns.map((t): TestCase => [t, PrettyFormats.GAMMA, '916 123-4567']),
            ...testMsisdns.map((t): TestCase => [t, PrettyFormats.DELTA, '(916) 123-4567']),
            ...testMsisdns.map((t): TestCase => [t, PrettyFormats.EPSILON, '+7 (916) 123-4567']),
            ...testMsisdns.map((t): TestCase => [t, PrettyFormats.ZETA, '+7 (916) 123-45-67']),
            ...testMsisdns.map((t): TestCase => [t, PrettyFormats.ETA, '+7 916 123-45-67']),
            ...testMsisdns.map((t): TestCase => [t, undefined, '+7 (916) 123-45-67']),
            ...testMsisdns.map((t): TestCase => [t, 'alpha', '9161234567']),
            ...testMsisdns.map((t): TestCase => [t, 'unknown-format', '+7 (916) 123-45-67']),
        ];

        it.concurrent.each(testTable)('pretty(%s, %p) should return %s',
            (msisdn, format, expected) => {
                expect(pretty(msisdn, format)).toEqual(expected);
            }
        );
    });

    describe('pretty incorrect cases', () => {
        const invalidMsisdns = [
            true,
            {},
            { number: 79060523777 },
        ];

        type TestCase = [any, any];
        const testTable: TestCase[] = invalidMsisdns.map((t): TestCase => ([t, t]));

        it.concurrent.each(testTable)('pretty(%s) should return %s',
            (msisdn, expected) => {
                expect(pretty(msisdn)).toEqual(expected);
            }
        );
    });

    describe('clean correct cases', () => {
        const testMsisdns = [
            '89161 23 45-67',
            '+7 916 123-4567',
            '79161234567',
            '9161234567',
            '916 123-4567',
            '(916) 123-4567',
            '+7 (916) 123-4567',
        ];

        type TestCase = [string, boolean | undefined, string];
        const testTable: TestCase[] = [
            ...testMsisdns.map((t): TestCase => [t, undefined, '79161234567']),
            ...testMsisdns.map((t): TestCase => [t, false, '79161234567']),
            ...testMsisdns.map((t): TestCase => [t, true, '9161234567']),
        ];

        it.concurrent.each(testTable)('clean(%p, %p) should return %s',
            (msisdn, removeLeadingSeven, expected) => {
                expect(clean(msisdn, removeLeadingSeven)).toEqual(expected);
            }
        );
    });

    describe('clean incorrect cases', () => {
        const invalidMsisdns = [
            89060523777,
            79060523777,
            9060523777,
            {},
            true,
        ];

        type TestCase = [any, boolean | undefined, null];
        const testTable: TestCase[] = [
            ...invalidMsisdns.map((t): TestCase => [t, undefined, null]),
            ...invalidMsisdns.map((t): TestCase => [t, false, null]),
            ...invalidMsisdns.map((t): TestCase => [t, true, null]),
        ];

        it.concurrent.each(testTable)('clean(%s, %p) should return %s',
            (msisdn, removeLeadingSeven, expected) => {
                expect(clean(msisdn, removeLeadingSeven)).toEqual(expected);
            }
        );
    });

    describe('pretty customPattern cases', () => {
        const msisdn = '88161234567';
        const expected = '+7 816 123-45-67';

        it('pretty 88161234567 with custom pattern should be return +7 816 123-45-67', ()=> {
            const customPattern = /^7?[4689]\d{9}$/;

            expect(pretty(msisdn, PrettyFormats.ETA, customPattern)).toEqual(expected);
        })
        
        it('not should be pretty', ()=> {
            const customPattern = /^7?[469]\d{9}$/;

            expect(pretty(msisdn, PrettyFormats.ETA, customPattern)).toEqual(msisdn);
        })
           
    });

    describe('clean customPattern cases', () => {
        const msisdn = '8 816 123 45 67';
        const expected = '8161234567';

        it('call with custom pattern', ()=> {
            const customPattern = /^7?[4689]\d{9}$/;

            expect(clean(msisdn, true, customPattern)).toEqual(expected);
        })
        
        it('call without custom pattern', ()=> {
            expect(clean(msisdn, false)).not.toEqual(expected);
        })
           
    });
});


rollup.config.ts
/* eslint-disable import/no-extraneous-dependencies */
import babel from '@rollup/plugin-babel';
import { defineConfig } from 'rollup';
import dts from 'rollup-plugin-dts';

export default defineConfig([
    {
        input: 'src/index.ts',
        output: [
            {
                file: 'dist/index.es.js',
                format: 'es',
            },
            {
                file: 'dist/index.cjs.js',
                format: 'cjs',
            },
        ],
        plugins: [
            babel({
                babelHelpers: 'bundled',
                exclude: 'node_modules/**',
                extensions: ['.ts'],
            })
        ]
    },
    {
        input: 'src/index.ts',
        output: [
            {
                file: 'dist/index.d.ts',
                format: 'es',
            },
        ],
        plugins: [dts()],
    },
]);


src/index.ts
const patterns = {
    validMsisdn: /^7?[469]\d{9}$/,
    notDigitsNorPlus: /[^\d+]+/g,
    cleanMsisdnParts: /^(?:\+?7|8)?(\d{3})(\d{3})(\d{2})(\d{2})$/,
};

/**
 * Msisdn formats.
 *
 * ALPHA   -> 9161234567
 * BETA    -> 79161234567
 * GAMMA   -> 916 123-4567
 * DELTA   -> (916) 123-4567
 * EPSILON -> +7 (916) 123-4567
 * ZETA    -> +7 (916) 123-45-67
 * ETA     -> +7 916 123-45-67
*/

export enum PrettyFormats {
    ALPHA = 'alpha',
    BETA = 'beta',
    GAMMA = 'gamma',
    DELTA = 'delta',
    EPSILON = 'epsilon',
    ZETA = 'zeta',
    ETA = 'eta',
}

const prettyFormatsRegExp = {
    [PrettyFormats.ALPHA]: '$1$2$3$4',
    [PrettyFormats.BETA]: '7$1$2$3$4',
    [PrettyFormats.GAMMA]: '$1 $2-$3$4',
    [PrettyFormats.DELTA]: '($1) $2-$3$4',
    [PrettyFormats.EPSILON]: '+7 ($1) $2-$3$4',
    [PrettyFormats.ZETA]: '+7 ($1) $2-$3-$4',
    [PrettyFormats.ETA]: '+7 $1 $2-$3-$4',
};

/**
 * Cleans msisdn.
 *
 * @param {string} msisdn msisdn to clean
 * @param {boolean} [removeLeadingSeven=false] Remove leading 7
 * @param {RegExp} customPattern custom pattern for validation msisdn
 * @returns {?string} cleaned msisdn
 */
export const clean = (msisdn: string, removeLeadingSeven = false, customPattern?: RegExp): string | null => {
    if (typeof msisdn !== 'string') {
        return null;
    }

    const cleaned = msisdn.replace(patterns.notDigitsNorPlus, '');
    const resultMsisdn = cleaned.replace(
        patterns.cleanMsisdnParts,
        removeLeadingSeven ? prettyFormatsRegExp[PrettyFormats.ALPHA] : prettyFormatsRegExp[PrettyFormats.BETA],
    );

    if (customPattern) {
        return customPattern.test(resultMsisdn) ? resultMsisdn : null;
    }

    return patterns.validMsisdn.test(resultMsisdn) ? resultMsisdn : null;
};

/**
 * Makes msisdn pretty.
 *
 * @param {number|string} msisdn msisdn in string or number.
 * @param {PrettyFormats} [format=PrettyFormats.ZETA] format
 * @param {RegExp} customPattern custom pattern for validation msisdn
 * @returns {string} msisdn string in pretty format
 */
export const pretty = (msisdn: string | number, format: PrettyFormats | string = PrettyFormats.ZETA, customPattern?: RegExp): string => {
    let msisdnStr = msisdn;
    if (typeof msisdnStr === 'number') {
        msisdnStr = String(msisdn);
    }

    const cleaned = clean(msisdnStr, false, customPattern);
    if (!cleaned) {
        return msisdnStr;
    }

    const unsafeFormat: PrettyFormats = format as PrettyFormats;
    const safeFormat: PrettyFormats = prettyFormatsRegExp[unsafeFormat] ? unsafeFormat : PrettyFormats.ZETA;

    return cleaned.replace(patterns.cleanMsisdnParts, prettyFormatsRegExp[safeFormat]);
};
